### PR TITLE
Use the function (closure) form of tweens

### PR DESCRIPTION
**Pull request by seanh**
Use the simpler function / closure form for implementing a Pyramid tween, rather than the class form.

The class form is a little dodgy:

> You should avoid mutating any state on the tween instance. The tween is invoked once per request and any shared mutable state needs to be carefully handled to avoid any race conditions.
> 
> The closure style performs slightly better and enables you to conditionally omit the tween from the request processing pipeline (see the following timing tween example), whereas the class style makes it easier to have shared mutable state and allows subclassing.
>
> https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hooks.html?highlight=add_tween#creating-a-tween

Also split things up into four separate, small tweens that each do one thing rather than two big tweens that do multiple things. This makes a couple of things simpler:

* We can conditionally add or not add each tween depending on which features are enabled by the settings. So the setting-based logic can now all live in one place `includeme()` rather than being split between the `includeme()` and the tweens themselves.

* The tweens themselves are now dirt simple and each do only one job. No more settings logic in the tweens. One tween per feature

* It's no longer necessary for `includeme()` to add the `SanitySettings` object to the registry, because the tweens no longer need it

* `SanitySettings.ingress_required` and `SanitySettings.egress_required` are no longer needed

### CODE REVIEW

**Comment by seanh** [comment]
This actually seems to break the extension. Now when I do `http 'http://127.0.0.1:5000/?f%FC=123'` to h it's 500ing instead of returning Bad Request. The extension _is_ raising `InvalidQueryString` though

File: src/pyramid_sanity/__init__.py [hunk 1]
```diff
@@ -1,7 +1,11 @@
 from pyramid.tweens import MAIN
 
-from pyramid_sanity._egress import EgressTweenFactory
-from pyramid_sanity._ingress import IngressTweenFactory
+from pyramid_sanity._egress import ascii_safe_redirects_tween_factory
+from pyramid_sanity._ingress import (
+    invalid_form_tween_factory,
+    invalid_path_info_tween_factory,
+    invalid_query_string_tween_factory,
+)
```
Comment (seanh) [inline src/pyramid_sanity/__init__.py:8]: https://github.com/hypothesis/pyramid-sanity/pull/3/ gets rid of these imports

File: src/pyramid_sanity/__init__.py [hunk 2]
```diff
@@ -11,24 +15,26 @@ def includeme(config):
     """Initialize this extension."""
     settings = SanitySettings.from_pyramid_settings(config.registry.settings)
 
-    # Put the settings into the registry so that other parts of the
-    # pyramid_sanity code can access them.
-    config.registry.settings["pyramid_sanity"] = settings
```
Comment (seanh) [inline src/pyramid_sanity/__init__.py:16]: This is no longer needed

File: src/pyramid_sanity/__init__.py [hunk 3]
```diff
@@ -11,24 +15,26 @@ def includeme(config):
...
-        # Other tweens that get added later might get added above this tween, but
-        # it's up to the app to manage that.
-        config.add_tween("pyramid_sanity.IngressTweenFactory")
+    if settings.check_params:
+        config.add_tween("pyramid_sanity.invalid_query_string_tween_factory")
+
+    if settings.check_path:
+        config.add_tween("pyramid_sanity.invalid_path_info_tween_factory")
 
-    if settings.egress_required:
```
Comment (seanh) [inline src/pyramid_sanity/__init__.py:27]: `ingress_required` and `egress_required` are no longer needed

File: tests/unit/pyramid_sanity/conftest.py [hunk 4]
```diff
@@ -3,15 +3,15 @@
...
 from pyramid.response import Response
+from pyramid.testing import DummyRequest
 
 from pyramid_sanity._settings import SanitySettings
 
 
 @pytest.fixture
 def pyramid_request():
-    return create_autospec(Request, instance=True)
+    return DummyRequest()
```
Comment (seanh) [inline tests/unit/pyramid_sanity/conftest.py:14]: The tests were easier to write with a `DummyRequest` than with a mock. Also see this old comment about why mocking `Request` is a bad idea: https://github.com/hypothesis/h/pull/5274#discussion_r215934743

**Review by jon-betts — APPROVED** [review]
I think we ironed out all the problems

**Pull request merged**